lib/cocoapods-core/cdn_source.rb

```diff
@@ -369,104 +386,112 @@ def download_and_save_with_retries_async(partial_url, file_remote_url, etag, ret
+        # We need to update the file modification date, as it is later used for freshness
+        # optimization. See #initialize for more information.
+        FileUtils.touch path
+      when 200
+        File.open(path, 'w') do |f|
+          encoding = response.headers['content-encoding'].to_s
+          if encoding.present?
+            case encoding
+            when 'gzip'
+              body = Zlib::GzipReader.wrap(StringIO.new(body), &:read)
+            else
+              raise Informative, "CDN: #{name} URL couldn't be saved: #{file_remote_url} Content encoding: #{response.headers['content-encoding']}"
             end
           end
+
+          f.write(body&.force_encoding('UTF-8'))
+        end
+
+        etag_new = response.headers['etag']
+        debug "CDN: #{name} Relative path downloaded: #{partial_url}, save ETag: #{etag_new}"
+        File.open(etag_path, 'w') { |f| f.write(etag_new) } unless etag_new.nil?
+      when 404
+        debug "CDN: #{name} Relative path couldn't be downloaded: #{partial_url} Response: #{response.status}"
+        nil
+      when 502, 503, 504
+        # Retryable HTTP errors, usually related to server overloading
+        message = "CDN: #{name} URL couldn't be downloaded: #{file_remote_url} Response: #{response.status} #{body}"
+        if retries <= 1
+          raise Informative, message
         else
-          raise Informative, "CDN: #{name} URL couldn't be downloaded: #{file_remote_url} Response: #{response.response_code} #{response.response_body}"
+          debug message + ", retries: #{retries - 1}"
+          make_sleep backoff_time(retries)
+          download_and_save_with_retries_async(partial_url, file_remote_url, etag, retries - 1)
         end
+      else
+        raise Informative, "CDN: #{name} URL couldn't be downloaded: #{file_remote_url} Response: #{response.status} #{body}"
+      end
+    end
+
+    def make_download(file_remote_url, etag = nil)
+      headers = [
+        %w[Accept-Encoding gzip],
+      ]
+
+      unless etag.nil?
+        headers << ['If-None-Match', etag]
       end
 
-      # Calling `Future#run` flattens the chained futures created by retries or redirects
-      #
-      # Does not, in fact, run the task - that is already happening in Hydra at this point
-      download_task.run
+      begin
+        netrc_info = Netrc.read
+        netrc_host = URI.parse(file_remote_url).host
+        credentials = netrc_info[netrc_host]
+        if credentials
+          user, pass = credentials
+          headers << ['Authorization', Protocol::HTTP::Header::Authorization.basic(user, pass)]
+        end
+      rescue Netrc::Error => e
+        raise Informative, "CDN: #{e.message}"
+      end
+
+      http_client.get file_remote_url, headers
     end
 
-    def exponential_backoff_async(retries)
-      sleep_async(backoff_time(retries))
+    def make_sleep(retries)
+      Async::Task.current.sleep backoff_time(retries)
```

Comment:
@jasl There's a problem here that wasn't caught by the tests: 

Method is named `#make_sleep` and accepts param `retries`, but all the calls are passing time in seconds, thus `#backoff_time` is happening twice, incorrectly. 

Fixes:
* rename to `#sleep_async`
* rename param to `seconds`
* remove inner call to `#backoff_time`

Reply:
my bad, applied your suggests.